### ISSUE TITLE
[HOLD for payment 2024-10-29] [HOLD for payment 2024-10-25] [$250] The waypoint editor opens as a Not Found page

### ISSUE THREAD

**Issue by m-natarajan** [Expensify/App#50961]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:**
**Reproducible in staging?:** need reproduction
**Reproducible in production?:** needs reproduction
**If this was caught on HybridApp, is this reproducible on New Expensify Standalone?:**
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Expensify/Expensify Issue URL:**
**Issue reported by:** @paultsimura 
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1729104627493469

## Action Performed:

1. Go to self-dm
2. Click "+" -> Track Expense (make sure the first tab that opens is the Distance request – if not, select it, close the RHP and start the Track Expense flow again)
3. Click the first waypoint

## Expected Result:
the waypoint editor opens
## Actual Result:
 the "Not found" page opens
## Workaround:
unknown
## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android: Standalone
- [ ] Android: HybridApp
- [ ] Android: mWeb Chrome
- [ ] iOS: Standalone
- [ ] iOS: HybridApp
- [ ] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [ ] MacOS: Desktop

## Screenshots/Videos

https://github.com/user-attachments/assets/04b59d1b-3661-44ba-83b6-8127d3558356

<details>
  <summary>Add any screenshot/video evidence</summary>

  
</details>

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~021846693147955823744</li>
        <li>Upwork Job ID: 1846693147955823744</li>
        <li>Last Price Increase: 2024-10-16</li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @trjExpensify</details>

**Comment by MelvinBot** [comment]
This has been labelled "Needs Reproduction". Follow the steps here: https://stackoverflowteams.com/c/expensify/questions/16989

**Comment by trjExpensify** [comment]
Hm, I can't really triage this if it's dev only. @paultsimura adding you as you reported it. 👍

**Comment by mkzie2** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
the "Not found" page opens

### What is the root cause of that problem?

Regression from https://github.com/Expensify/App/pull/50772

Not found page in waypoint page would show if there's no waypoints from the draft transaction.

In this case when we get the correct request type:

https://github.com/Expensify/App/blob/a1d92ef9feae7fd81c35956164a13d9ffad7fb02/src/pages/iou/request/IOURequestStartPage.tsx#L60

to init the draft transaction where we initialize the optimistic waypoints:

https://github.com/Expensify/App/blob/a1d92ef9feae7fd81c35956164a13d9ffad7fb02/src/pages/iou/request/IOURequestStartPage.tsx#L68

https://github.com/Expensify/App/blob/3a1d1a8fa0968da093b938aaf650e4f92b330709/src/libs/actions/IOU.ts#L341-L351

At this time the `transactionDraft` is `null` causing the `transactionRequestType` to be `manual` while it should be `distance`.

However, before https://github.com/Expensify/App/pull/50772, we attached a screen listener that triggers `onTabSelected` and then call `resetIOUTypeIfChanged` which eventually initialize the draft transaction with the correct request type (`distance`):

https://github.com/mkzie2/App/blob/a1d92ef9feae7fd81c35956164a13d9ffad7fb02/src/pages/iou/request/IOURequestStartPage.tsx#L85

So this works fine on staging and production.

But after https://github.com/Expensify/App/pull/50772, `onTabSelected` is only triggered when we manually select a new tab:

https://github.com/Expensify/App/blob/286212cc4f23b94426dca0bc28a039f61ca5c693/src/libs/Navigation/OnyxTabNavigator.tsx#L127-L129

This causes a problem when we previously select a tab, then close the RHP and reopen the RHP, the `selectedTab` is still the previously selected tab and thus `onTabSelected` is not triggered > `resetIOUTypeIfChanged` is not called > draft transaction is not init with the correct request type/tab.

### What changes do you think we should make in order to solve the problem?


We should not check the request type barely based on transaction. It is `null` almost the time since we `clearMoneyRequest` before starting any request:

https://github.com/Expensify/App/blob/a1d92ef9feae7fd81c35956164a13d9ffad7fb02/src/pages/iou/request/IOURequestStartPage.tsx#L60

We can use `selectedTab` if `transaction` is `undefined`:

```javascript
const transactionRequestType = useMemo(() => !isEmptyObject(transaction) ? TransactionUtils.getRequestType(transaction) : selectedTab, [transaction, selectedTab]);
```

https://github.com/Expensify/App/blob/a1d92ef9feae7fd81c35956164a13d9ffad7fb02/src/pages/iou/request/IOURequestStartPage.tsx#L63-L69

```javascript
useEffect(() => {
    if (transaction?.reportID === reportID || isLoadingSelectedTab) {
        return;
    }
    IOU.initMoneyRequest(reportID, policy, isFromGlobalCreate, transactionRequestType);
}, [transaction, policy, reportID, iouType, isFromGlobalCreate, isLoadingSelectedTab]);
```

### What alternative solutions did you explore? (Optional)

NA

**Comment by mkzie2** [comment]
@trjExpensify I think this should be a deploy blocker as a regression from https://github.com/Expensify/App/pull/50772 which will be deployed soon.

**Comment by trjExpensify** [comment]
Noted: https://github.com/Expensify/App/pull/50772#issuecomment-2418129153

**Comment by bernhardoj** [comment]
I left a comment [here](https://github.com/Expensify/App/pull/50772#issuecomment-2418437533) suggesting to fix this as a separate issue.

**Comment by trjExpensify** [comment]
Yeah, I don't agree that it's a separate issue. That PR now causes users to hit an error page when they open the waypoints editor. We shouldn't send that to prod until it's fixed.

**Comment by trjExpensify** [comment]
>@trjExpensify what is your take here? Do you think we should let the offending PR author @bernhardoj fix this?

Yes, I think it should be treated as a regression personally.

**Comment by bernhardoj** [comment]
> We shouldn't send that to prod until it's fixed.

I agree that we should fix it before reaching prod, but the reason I personally don't think it counts as a regression is, that I or @rayane-djouah wouldn't expect that the money request initialization is not working correctly. It's a hidden bug and never works which is only discovered after the PR. There have been a few cases like this in the past.

In the past, I usually do the follow-up, but it doesn't count as a regression. I suggest treating it as a separate issue since @mkzie2 has already proposed a working solution. I can open the PR but I will borrow the solution above 😅.

**Comment by mkzie2** [comment]
This makes the Submit Scan/Distance expense flow completely malfunctioned and should be fixed with urgency. I can raise PR immediately if needed.

**Comment by bernhardoj** [comment]
Ok, looks like the init issue causes other flows (https://github.com/Expensify/App/issues/51007 & https://github.com/Expensify/App/issues/51029) too which isn't covered by the proposal, so I'm going to do the follow-up PR. I'll explain all the root causes in the PR.

**Comment by trjExpensify** [comment]
@flodnv is already assigned here, removing Ben.

**Comment by marcaaron** [comment]
CP'd the fix I will ask for a re-test when it succeeds.

**Comment by kavimuru** [comment]
Bug is fixed.

https://github.com/user-attachments/assets/0cd6a8cc-7b4e-4ee6-a48c-a78158885956

**Comment by trjExpensify** [comment]
👋 Checklist time please, @paultsimura!

**Comment by trjExpensify** [comment]
Ah cool, okay. Closing it out then!
